### PR TITLE
Local/Whisper/Emote/Radio/LOOC/DeadChat Chat Refactor

### PR DESCRIPTION
**Pull request by FairlySadPanda**
## About the PR
Refactors the following types of chat to use bespoke code paths rather than being smushed together inside ChatSystem:

LocalChat (normal, hidden-from-log and subtle)
Whispering (close, far and out-of-eyesight)
Emoting
Dead Chat
LOOC
All radio channels, including announcements.

This does not include much work on:

OOC
Admin Chat
System messages

## Why / Balance
The old code is shit.

## Technical details
_breathes in_

Chat has been restructured to use distinct code paths. For example Local Chat (as in, just talking) has a ClientLocalChatSystem, LocalChat events, a ServerLocalChatSystem, etc.

Formatting of chat messages has moved client-side. That means deciding how to display a chat message that has been received is now largely (although not entirely at the moment, because verbs and their associated fonts, boldness and size are still pulled from proto and sent from the server) a job of the somewhat enlarged ChatUIController class.

The pattern for sending a chat message is done via an attempt event from the client. The event attempt will be followed by a failure event back to the attempter (which will then pop up the reason for the failure as a popup) or will go through and issue an EntityXedEvent of the type of chat they did. There can be a few Xed events - whispering has four, for example.

I've broken out as much code as I can touch into helper managers, and attempted to avoid any regressions. However, there are some consumers of this system (cameras, mostly) that might have problems.

Note that where I've edited YAML I've also had an automated linter format it, hence the inflated line count. Remember you can disable showing whitespace changes when reviewing PRs.

## Media
- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
This totally shorts out the current local chat, whisper and emote handling and puts it into its novel code. If you merge this, may God have mercy on your souls. This needs a bunch of feedback and testing.

Specifically right now this has outright removed the "don't show this in local chat" etc code. This probably has a knock-on to chat usefulness, as automata like medibots now appear.

Adding this back in is probably not going to be too bad, but it's a bit of extra work we don't need to do to get this PR reviewed.

**Changelog**
:cl:
- fix: Starting/stopping a tail wag now displays an emote pop-up, but it doesn't display in the chat log.
- fix: Starting/stopping wagging your tail is on a one-second cooldown to prevent spamming.
- add: Handheld radios now light up when switched on.
- tweak: Shouting and whispering has tuned text formatting. Shouting makes your text bigger!
- tweak: Animals and other non-named mobs have had their number removed. Hamlet is not a number!

### CODE REVIEW

**Comment by FairlySadPanda** [comment]
DO NOT MERGE THIS PR.

**Comment by mirrorcult** [comment]
godspeed soldier

**Comment by deltanedas** [comment]
sisyphus pushing shitcode up a mountain :trollface:

**Comment by nikthechampiongr** [comment]
Holy shit.

File: Content.Server/Chat/V2/ServerLocalChatSystem.cs
Comment (FairlySadPanda) [inline Content.Server/Chat/V2/ServerLocalChatSystem.cs:82]: Out of date comments

**Comment by Geekyhobo** [comment]
Holy shit

**Comment by FairlySadPanda** [comment]
TODOs

1. Migrate Whisper and Emote over.
2. Get the admins to tell me what features they want out of this system (i.e. better chat message moderation)

**Comment by FairlySadPanda** [comment]
This is a tonne of work and a huge set of changes; there's basically two clumps of work.

1. New code paths for dealing with local, whisper and emote chat.
2. Orchestration to make other code work with them.

This was originally going to be an additive-only change, but there's a number of server-side systems that directly call ChatSystem TrySendInGameICMessage and similar functions and this meant making breaking changes instead; it was unfeasible to do a feature flag for this code without the flag being in at least two dozen bits of code.

Quite a bit of this code is still closely based on the origin code, usually from ChatSystem. This will be an avenue for improvement as this PR gets reviewed, because I think there's a lot of cleanup that can still be done.

**Comment by FairlySadPanda** [comment]
TODOS

1, Get rate limiting confirmed working properly
2. Work on chat word filtration

**Comment by VasilisThePikachu** [comment]
Under request of PR owner

**Comment by FairlySadPanda** [comment]
_sigh_

Think this is ready for review. There's still a potential for weirdness in harder-to-test scenarios like handheld radio, off-grid comms...

**Comment by FairlySadPanda** [comment]
My continuing to-do will be automated chat filtration; this will need some file IO work in a filtration manager.

**Comment by FairlySadPanda** [comment]
Automated chat filtration is done and ready to examine.

This PR is now 100% feature complete and will receive no further notable changes other than to resolve merge conflicts.

@Jezithyr @DrSmugleaf @Chief-Engineer @EmoGarbage404 I'd appreciate a review, or please hand off the review responsibility to folks with time, or please let me know if it'll be a while before it can be reviewed. Also LMK if there's anything I can do to make reviewing easier. This is a lot of changes so, uh, I can help where I can.

File: Content.Client/Chat/V2/ClientEmoteSystem.cs
Comment (Jezithyr) [inline Content.Client/Chat/V2/ClientEmoteSystem.cs:19]: Needs a loc-string

File: Content.Client/Chat/V2/ClientLocalChatSystem.cs
Comment (Jezithyr) [inline Content.Client/Chat/V2/ClientLocalChatSystem.cs:19]: needs a loc-string

File: Content.Client/Chat/V2/ClientRadioSystem.cs
Comment (Jezithyr) [inline Content.Client/Chat/V2/ClientRadioSystem.cs:22]: needs a loc-string

File: Content.Client/Chat/V2/ClientRadioSystem.cs
Comment (Jezithyr) [inline Content.Client/Chat/V2/ClientRadioSystem.cs:30]: needs a loc-string

File: Content.Client/Chat/V2/ClientWhisperSystem.cs
Comment (Jezithyr) [inline Content.Client/Chat/V2/ClientWhisperSystem.cs:16]: Remove this if it is no longer being used?

File: Content.Server/Chat/Managers/ChatSanitizationManager.cs
Comment (Jezithyr) [inline Content.Server/Chat/Managers/ChatSanitizationManager.cs:12]: This should definately be moved to yaml prototypes in a future PR by someone, especially since it would allow for multiple different filter lists to be used at the same time.

**Review by Jezithyr — COMMENTED** [review]
The rest of the changes look good to me. I'm going to ask another maint to take a look as a second opinion but this is a massive improvement over the status quo.

**Review by EmoGarbage404 — CHANGES_REQUESTED** [review]
your IDE ran indentation formatting changes on every single YAML file, which completely bloats the diff and makes this a conflict magnet.

**Comment by FairlySadPanda** [comment]
> your IDE ran indentation formatting changes on every single YAML file, which completely bloats the diff and makes this a conflict magnet.

Obviously the solution is to lint every single YAML file ;)

Will fix. For the time being you can disable viewing whitespace to see the relevant changes.

**Comment by MilenVolf** [comment]
CL should describe something that players can see in game. Players don't need to know about code refactor if it doesn't change anything in gameplay.

**Comment by FairlySadPanda** [comment]
> CL should describe something that players can see in game. Players don't need to know about code refactor if it doesn't change anything in gameplay.

Good point. I can only think of four actual player-faxing changes. I have added them.

**Comment by MilenVolf** [comment]
> > CL should describe something that players can see in game. Players don't need to know about code refactor if it doesn't change anything in gameplay.
> 
> Good point. I can only think of four actual player-faxing changes. I have added them.

Fix CL format from `-fix` to `- fix`, just add space to every change after hyphen

File: Resources/Prototypes/Entities/Structures/Machines/anomaly_equipment.yml
Comment (FairlySadPanda) [inline Resources/Prototypes/Entities/Structures/Machines/anomaly_equipment.yml:80]: Missed removing one lint change :theendtimes:

**Comment by FairlySadPanda** [comment]
@EmoGarbage404 Had to brutalize the PR history but you'll note the changes are now reduced in size

**Comment by mirrorcult** [comment]
im dubious on the actual value of this refactor. i dont think "similar methods of chatting shared the same functions to reduce duplication" was the problem with the current chat system and i dont think duplicating at least 1.5k lines of identical code across like 8 different entitysystems for every chat channel is an improvement on that front

in my opinion the main issues with the current chat system are generally as follows:
1) the APIs were hard to understand due to having new optional/required args bolted on and the order not making a lot of sense--was very easy to pass bools to the wrong args and fuck shit up for instance
2) message wrapping and formatting with font coloring/bolding/etc is handled really terribly and clientside handling of messages was janky and not conducive to further improvements (though idk the exact specifics that much)
3) control flow was very confusing and it made it hard to understand where stuff needed to go in the pipeline when adding new features (this was the worst problem probably)
4) all transformation was handled serverside and destructively, so no support for clientside languages or reconstructing original messages
5) handling potential recipients was very confusing and annoying and generally not very easily configurable

on **1**, this PR seems to do a better job by handling a lot of checks only in the network event handler instead of in the actual API, but simultaneously it also makes the API confusing in other ways--there isnt a way to check 'canemote/looc/chat/whisper' because some of the checks are located in the event handler/a "try(etc)" function/just in the actual sendemote/looc/etc function and its inconsistent where it goes. e.g. sendemote will send even if the entity passed in isnt emoteable which is confusing, but specifically tryemote will fail at that, but it wont fail with any of the other checks like message length or rate limiting. in looc, sendlooc will fail in some situations rather than force a looc--i guess because there isnt really any delineation or standardization between what gets checked for ic chat (local/whisper/etc) / ooc chat (looc/dead/etc) like there was in the previous system

as for other sections of the API like the chat events.. uhh... this is really bad still

image

on **2**, for sure for sure an improvement on the old system from what i can tell. not much else to say other than i think the improvements here could stand on their own without most of the other serverside changes, and also that the improvements dont really go far enough (though this is commented on in relevant code paths), still a decent amount of jank with it

on **3**, this PR is somewhat an improvement, but generally at the cost of duplicating code paths, sometimes literally just copy-pasted especially between stuff like local/whisper which ideally share like 90% of their code. in the future if this was merged i would expect it to become techdebt as people forget to add support to every code path/do things slightly differently and they diverge in behavior

on **4**, this PR seems to not have improved the situation at all--message transformation looks to be handled destructively in the same way as previously

on **5**, same as previously as well. there doesnt look to be any way for callers to customize this behavior (or hooks to customize the behavior like with events or anything) and its mostly just handled bespoke and copy-pasted by every individual system

there are some good changes here (mostly in point 2 mentioned here) but overall i am primarily very very confused by the decision to split everything into copy pasted bespoke code paths. for 4 and 5 its understandable if you just feel that its scope creep but i dont see a reason to rewrite the entire server chat system if its not going to consider things like this as fundamental improvements that should be considered from the outset

**Comment by mirrorcult** [comment]
overall this just feels like 2 steps forward 1 step back for no discernible reason, when to have a chat system that actually meets all our requirements needs like 5-6 steps forward--from the start i've been saying this has really needed to be a fully holistic refactor that takes into account the features we actually want out to be possible out of a refactor

**Comment by metalgearsloth** [comment]
I mostly agree with mirror. I think point 3 is the main thing that needs entirely ripping out of chat at some point before we try and do anything more with it as this refactor doesn't touch chatmanager or chatsystem at all.

> control flow was very confusing and it made it hard to understand where stuff needed to go in the pipeline when adding new features (this was the worst problem probably)

**Comment by metalgearsloth** [comment]
Here's my high-level list of what still needs doing before firefox explodes again. I'm also going to include some specific stuff I reviewed because github lags to shit for me when I open the PR:
- The copy-paste between whisper / radio / every other system is really bad e.g. 'Nice try bozo' being copied at least 6 times, same with rate limiting and the likes. These should just be chatsystem partials.
- If the chat messages are tied to the player's entity then just use the player's entity.
- Validating messages needs to be done in a centralised area rather than being in message handlers (this is so we can handle stuff like non-player chat properly).
- Stuff like "shouldCapitalizeTheWordI" should be pulled out entirely and not baked in. Right now the systems still have chat transformation too entrenched and it's going to make any implementation of languages a hassle.
- Managers are strictly for anything that needs to run out of the game sim (e.g. on the main menu) so should be used sparingly. Most of the new ones you added should just be systems (especially if they have an entitymanager dependency).
- Some of the stuff like station announcement audio should really be split out into their own methods rather than copied between each one.
- 90% of ChatManager needs moving to ChatSystem i.e. everything but OOC and admin I think, as most of it is tied to the game sim and has no business being in it. Frankly 90% of it needs to be bulldozed because wrappedmessage is awful but yeah.
- It would be nice if some of the control flow were documented better, e.g. sendemotemessage randomly terminates at several points and I have to open other methods up to see why.
- RadioSystem should be using events for stuff like removing channels e.g. if a borg equips a headset it will nuke any implicit radio channels they may have when it's removed.
- Some of the data between the different chat events is also copied and would be better to be more unified though I don't know the best way to do it, someone else may have a better idea (struct for the basic message + struct for markup? not sure).
- Inserting some kind of "transformchathere" or the likes for future archaeologists to handle languages too.

**Comment by FairlySadPanda** [comment]
All good feedback and good targets for further refactor. Can empathise with Github not liking large PRs.

Also note there's some bugs with this, probably due to the YAML cleanup. Sigh. Will squash those first and then translate feedback into actions and iterate, all that good stuff.

I'm also going to have to write up a full TDD for this, I suspect, as it sounds like getting signoff on technical approach trumps much further rework. If the scope of the project _requires_ a full rip-out-and-restart approach to even get merged then 💀 - I don't think that can be sensibly handled by a lone dev, especially as this code is not just core to Wizden but chat on most downstreams.

**Comment by FairlySadPanda** [comment]
OH just to be clear on scope

This is _specifically only for round-scoped chat_ at the moment - that's why ChatManager and some of ChatSystem has not been deleted. I agree that is where the shittiest code lies, but I had to draw the line somewhere otherwise I'd have ideaguy'd myself through the entire chat codebase and have something completely unmergeable, rather than, uh, mostly unmergeable

**Comment by FairlySadPanda** [comment]
> your IDE ran indentation formatting changes on every single YAML file, which completely bloats the diff and makes this a conflict magnet.

Could you clear this change request if it's alright now?

**Comment by FairlySadPanda** [comment]
@metalgearsloth @mirrorcult 

I've gone through and overhauled structure quite a bit to reduce code problems. You'll also hopefully note the overhaul to radio.

TODOs at the moment:
~~Get the new manager migrated to systems where possible (more idiomatic)~~
~~Migrate receive-transform-emit systems for round chat into partials of one overall system to limit code re-use (more idiomatic)
Find a nicer way of handling verb bloat in EntityLocallyChattedEvent (can I stop having to call events FooEvent? it stutters)~~
~~Get Borgs to stop whispering when beep-booping on Binary~~
~~Make sure intrinsic radio channels definitely absolutely do not get affected by gaining/losing non-intrinsic ones - I know I've sorted this but eh~~
Get PJB to shout at me about how shitty SimpleCensor is so I know how to optimize it, as it's currently doing a fucktonne of mallocs I can deffo avoid if I know the ways of real C# (as opposed to the zombified C# used for VRChat world development)
Make sure that handheld radios work
Make sure that cameras work
~~Fix whatever I broke when I removed those four and a half thousand lines of whitespace changes (thanks SonarLint!)~~
Make chat events more consistent: attempt, fail, success, local success, set of notices to recipients by recipient type

POTENTIAL
Migrate SimpleCensor to its own PR so folks who care about strings can shout at me separately 😉

TODOs for after this PR: 
get OOC/Admin/System refactored
add DOCUMENTED hooks for server-side text/recipient transformation
get frontend reception and translation of incoming chat events out of ChatUIController and into receive-transform-emit handlers
Look at verbs again because there's 100% bloat there
Get shorthand/smiley defined in YAML rather than in C#

**Comment by FairlySadPanda** [comment]
@PJB3005 y u conflic me :(

**Comment by PJB3005** [comment]
> @PJB3005 y u conflic me :(

The code must be cleansed...

Also please don't use rebases to fix conflicts if you can help it, it makes PR review harder. Your PR gets squashed in the end anyways so regular merge commits are preferred.

**Comment by FairlySadPanda** [comment]
> > @PJB3005 y u conflic me :(
> 
> The code must be cleansed...
> 
> Also please don't use rebases to fix conflicts if you can help it, it makes PR review harder. Your PR gets squashed in the end anyways so regular merge commits are preferred.

I get that, but merging in from head messes up my development process. I'll investigate seeing if I can squash pulls from master to avoid having a morbillion commits that are not mine in my commit history.

**Comment by FairlySadPanda** [comment]
@mirrorcult Take another look at the local chat code; huge amount of bloat reduction :)

**Comment by FairlySadPanda** [comment]
@Jezithyr Loc strings should all be in.

**Comment by FairlySadPanda** [comment]
Next up is making announcements nicer :>

**Comment by FairlySadPanda** [comment]
PR could be in a broken state atm, just did a bunch of work to kill off the V1 ChatSystem entirely. Means there's now some shitcode to deal with in V2 though.

The remaining ChatManager code has to remain as-is so I'm pretty much done here for clean-up. The rest of this PR is pretty much just handling feedback, hooks, etc.

**Comment by FairlySadPanda** [comment]
[4542a95](https://github.com/space-wizards/space-station-14/pull/24680/commits/4542a95015679d4f800f9353b772f22221d6e5e1) has a bunch of small fixes in it to make this actually compile and things like the comms console actually work.

This also ended up in a tangent where I ended up looking into the emote code and then got annoyed with "the mouse (567)..." being emitted for emotes. 

I've fixed things so that the text displays more correctly, and entity numbers are hidden when showing names. So if Hammy is squeaking, he's just Hammy, not Hammy (567), etc.

I will continue to bikeshed random tangential fixes inside this feature branch.

image

**Comment by FairlySadPanda** [comment]
Closing as stale; will individually introduce PRs per-file

**Pull request closed without merging**
